### PR TITLE
fix: manually retry ABORTED reads in transactions

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds manual retry to our transaction logic similar to what we have in the Web SDK: https://github.com/firebase/firebase-js-sdk/blob/c822e78b00dd3420dcc749beb2f09a947aa4a344/packages/firestore/src/core/transaction_runner.ts#L112

We only retry GRPC exceptions, and only failures with code ABORTED. This is because:
- we don't want to retry if the user throws an error or returns a rejected Promise
- GAPIC retries reads that fail with UNAVAILABLE, INTERNAL and DEADLINE_EXCEEDED under the hood

Fixes https://github.com/googleapis/nodejs-firestore/issues/827 (fingers crossed)

### CODE REVIEW

File: dev/test/transaction.ts [hunk 1]
```diff
@@ -106,7 +106,7 @@ function commit(
 
 function rollback(
   transaction?: Uint8Array | string,
-  err?: Error
+  error?: Error
```
Comment (thebrianchen) [inline dev/test/transaction.ts]: optional: change naming for `commit()` as well so they're all in sync ([link](https://github.com/googleapis/nodejs-firestore/blob/8f41b37fd766b18113401c7324e319b52b38cc96/dev/test/transaction.ts#L73))

File: dev/test/transaction.ts [hunk 2]
```diff
@@ -416,14 +424,51 @@ describe('failed transactions', () => {
...
+        await transaction.get(docRef);
+        return 'success';
+      },
+      begin('foo1'),
+      getDocument('foo1', retryableError),
+      rollback('foo1'),
+      begin('foo2', 'foo1'),
+      getDocument('foo2'),
+      commit('foo2')
+    ).then(red => {
```
Comment (thebrianchen) [inline dev/test/transaction.ts]: s/red/res

File: dev/test/transaction.ts [hunk 3]
```diff
@@ -416,14 +424,51 @@ describe('failed transactions', () => {
...
+
+  it("doesn't retry custom user exceptions in callback", () => {
     return expect(
       runTransaction(
         () => {
           return Promise.reject('request exception');
         },
         begin(),
-        rollback('foo')
+        rollback()
```
Comment (thebrianchen) [inline dev/test/transaction.ts]: should we make it consistent throughout [elsewhere](https://github.com/googleapis/nodejs-firestore/blob/8f41b37fd766b18113401c7324e319b52b38cc96/dev/test/transaction.ts#L399-L400)

File: dev/test/transaction.ts [hunk 4]
```diff
@@ -416,14 +424,51 @@ describe('failed transactions', () => {
...
+    const nonRetryableError = new GoogleError('Failed Precondition');
+    nonRetryableError.code = Status.FAILED_PRECONDITION;
+
+    return expect(
+      runTransaction(
+        async (transaction, docRef) => {
+          await transaction.get(docRef);
+          return 'failure';
+        },
+        begin('foo'),
```
Comment (thebrianchen) [inline dev/test/transaction.ts]: If 'foo' is default, do we need it here as well, or can it also be removed?

File: dev/src/transaction.ts [hunk 5]
```diff
@@ -552,3 +561,14 @@ function validateReadOptions(
     }
   }
 }
+
+function isRetryableTransactionError(error: Error): boolean {
+  if (error instanceof GoogleError || 'code' in error) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply dev/src/transaction.ts]: Yes, that is JavaScript functionality (see https://masteringjs.io/tutorials/fundamentals/hasownproperty for a comparison against `Object.hasOwnProperty`). I am doing it here because I am afraid that some of our dependencies might throw a GoogleError that is not actually a GoogleError by providing just a code and a message field.

File: dev/test/transaction.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply dev/test/transaction.ts]: Done. Good catch.

File: dev/test/transaction.ts (hunk 2 above)
Comment (schmidt-sebastian) [reply dev/test/transaction.ts]: Fixed (here and the test I copied it from)

File: dev/test/transaction.ts (hunk 4 above)
Comment (schmidt-sebastian) [reply dev/test/transaction.ts]: I left it as is here since `getDocument` uses the second argument. I prefer this over only specifying the transaction ID once or using `undefined` in its place.

File: dev/test/transaction.ts (hunk 3 above)
Comment (schmidt-sebastian) [reply dev/test/transaction.ts]: Fixed

File: dev/src/transaction.ts [hunk 6]
```diff
@@ -552,3 +561,14 @@ function validateReadOptions(
...
+    // In transactions, the backend returns code ABORTED for reads that fail
+    // with contention. These errors should be retried for both GoogleError
+    // and GoogleError-alike errors (in case the prototype hierarchy gets
```
Comment (novemberborn) [inline dev/src/transaction.ts]: Could this check for `error.name === 'GoogleError'` instead?

When implementing gRPC servers, it's not unthinkable that an ABORTED error is thrown from inside the transaction, with the expectation that it's returned to the client.

Comment (schmidt-sebastian) [reply dev/src/transaction.ts]: This should accept any object that has a "code" property. In the environment you are describing, will this property be stripped?

Comment (novemberborn) [reply dev/src/transaction.ts]: Let me rephrase.

There are many ways of wrongly implementing a transaction function, to do with side-effects, or forgetting to use the transaction itself to perform queries.

This new `ABORTED` handling adds more ways to get it wrong.

There is an assumption that all `GoogleError` instances came from transaction methods. That may not be the case if I use another GCP SDK within my function. I know, I shouldn't be, but somebody still might.

It also won't be the case if I call a Firestore method on a `DocumentReference` which then throws an `ABORTED` error.

These assumptions are then generalized by testing if the thrown error has a `.code = 10` property. I have a `ServiceError` object that is thrown with a gRPC response code. If for some reason I throw that error inside a transaction function, my transaction may be retried even if it succeeded. And again, sure, I suppose I shouldn't be throwing that error, but somebody still might.

A safer approach may be to register errors, thrown as part of the actual transaction, in a `WeakMap`. Then as they propagate through user code, when you catch them again you can check whether they're genuine transaction errors, and then check their code.

Comment (schmidt-sebastian) [reply dev/src/transaction.ts]: A couple things to note:
- A read outside of a transaction should never result in Code ABORTED. This error should only happen during a transactional read.
- If a user throws an error with `code: 10`, we will retry the transaction. In general, transaction functions should be written so they can be re-run, and re-running them more often than needed should not cause any harm.

I understand that we control the errors that we are catching here. We could tighten this check to make sure that the error is indeed a GoogleError, but for now I decided to err on the side of caution and also retry GoogleError-alike errors. This allows us to catch errors with code properties that were created by our dependencies before we converted everything to use the proper GoogleError type.
